2024-06-20-investigating-code-quality-from-pr-data.md

```diff
@@ -0,0 +1,75 @@
+
+These metrics are then saved to a Parquet file, which can be exported to a [ClickHouse](https://clickhouse.com/clickhouse) database and imported into [Apache Superset](https://superset.apache.org/) to create a dashboard which allows us to visualise and explore the data captured.
+
+### Let's see this in action
+To demonstrate our tool, we have used it to analyse 1000 PRs the main repositories for a number of different programming languages, namely [Python](https://github.com/python/cpython), [Rust](https://github.com/rust-lang/rust), [JDK](https://github.com/openjdk/jdk), [NodeJS](https://github.com/nodejs/node) and [.NET](https://github.com/microsoft/dotnet). We've then produced some charts displaying a range of the metrics we produce, with the aim of gaining an insight into the development of these languages.
+
+![jpg]({{ site.github.url }}/alaws/assets/code-quality/code-quality-analysis-date-dotnet.jpg
+ "Number of PRs opened per Month .NET")
+ ![jpg]({{ site.github.url }}/alaws/assets/code-quality/code-quality-analysis-date-combined.jpg
+ "Number of PRs opened per week")
+
+Firstly, a note on PR creation dates. Our tool gathers the 1000 most recent PRs from each repository. For CPython and Rust, these were all created between May and June 2024. In the case of NodeJS and Rust, the earliest PRs come from January and March 2024 respectively. In contrast, .NET, there are PRs from October 2014. Remarkably, there were 257 PRs opened in the CPython in the week commencing 20th May 2024.
+
+##### PR Duration and Review Time
+![png]({{ site.github.url }}/alaws/assets/code-quality/code-quality-analysis-pr-duration.jpg
+ "PR Duration")
+
+`PR Duration` is a measure of the time elapsed between a PR opened, and it ultimately being either merged or closed. However, it may be the case that a PR sits in review for a period of time, before the reviewer first looks at it. Similarly, once it is approved, there may be a time delay before it is merged. For this reason, we added the additional measures of `Time to First Review` and `Time From First Review to Last Commit`. This is perfectly illustrated by .NET, which has a noticeably high PR duration. However, if you raise a PR, you'd spend the majority of this time waiting for the first review, but would be likely to quickly merge/close your PR after the initial review.
+
+##### Contributors
+![png]({{ site.github.url }}/alaws/assets/code-quality/code-quality-analysis-contributions-per-user.jpg
+ "Number of Contributors Per User")
+
+Here, we define a contribution as having made a commit on a PR. The number of commits in that PR aren't considered. All languages seem to accept PRs from new contributors, but .NET and Rust seem to have some very seasoned contributors, with one person having worked on 161 and 157 PRs out of the last 1000.
```

Comment:
I am not sure what 'one person having worked on 161 and 157 PRS of the last 1000' means.